core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,162 @@
+    private final MeterIdPrefix meterIdPrefix;
+    private final DnsQuestion question;
+
+    private enum DnsExceptionTypes {
+        NX_DOMAIN_QUERY_FAILED_EXCEPTION,
+        CNAME_NOT_FOUND_QUERY_FAILED_EXCEPTION,
+        NO_MATCHING_RECORD_QUERY_FAILED_EXCEPTION,
+        UNRECOGNIZED_TYPE_QUERY_FAILED_EXCEPTION,
+        NAME_SERVERS_EXHAUSTED_EXCEPTION,
+        DNS_EXCEPTION,
+        DNS_TIMEOUT_EXCEPTION,
+        DNS_RESOLVER_TIMEOUT_EXCEPTION
+    }
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry,
+                                     DnsQuestion question, MeterIdPrefix prefix) {
+        this.meterRegistry = meterRegistry;
+        meterIdPrefix = prefix;
+        this.question = question;
+    }
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".written"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(SERVER_TAG, dnsServerAddress.getAddress().getHostAddress()))).increment();
+    }
+
+    @Override
+    public void queryCancelled(int queriesRemaining) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".cancelled"),
+                NAME_TAG, question.name()).increment();
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryRedirected(List<InetSocketAddress> nameServers) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".redirected"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(SERVERS_TAG,
+                        nameServers.stream().map(addr -> addr.getAddress().getHostAddress())
+                                .collect(Collectors.joining(","))))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryCNAMEd(DnsQuestion cnameQuestion) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".cnamed"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(CNAME_TAG, cnameQuestion.name()))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryNoAnswer(DnsResponseCode code) {
+        meterRegistry.counter(meterIdPrefix.name().concat(".noanswer"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(CODE_TAG, String.valueOf(code.intValue())))).increment();
+        return this;
+    }
+
+    @Override
+    public void queryFailed(Throwable cause) {
+        meterRegistry.counter(meterIdPrefix.name(),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(RESULT_TAG, "failure"),
+                        Tag.of(CAUSE_TAG, determineDNSExceptionTag(cause).name()))).increment();
```

Comment:
I guess almost all failed queries will end up getting `NAME_SERVERS_EXHAUSTED_EXCEPTION`, which means it's a kind of duplicate. Could we ignore it until we have a clear reason why we would want to record it?